### ISSUE TITLE
Attributes have incorrect span information

### ISSUE THREAD

**Issue by sgrif** [rust-lang/rust#47941]
Given the following struct:

```rust
#[derive(AsChangeset)]
#[table_name = "users"]
struct UserForm {
    id: i32,
    name: String,
}
```

the input to the custom derive will have incorrect span information for every token in the `table_name` attribute except for the pound. Every other token will have the span `Span { lo: BytePos(0), hi: BytePos(0), ctxt: #0 }`.

**Comment by dtolnay** [comment]
I was able to reproduce this with the following macro on rustc 1.25.0-nightly (56733bc9f 2018-02-01). The four tokens in the attribute, `[`...`]`, `table_name`, `=`, `"users"`, all have span `Span { lo: BytePos(0), hi: BytePos(0), ctxt: #0 }`.

#### src/main.rs

```rust
#![feature(proc_macro)]

extern crate repro;
use repro::AsChangeset;

#[derive(AsChangeset)]
#[table_name = "users"]
struct UserForm {
    id: i32,
    name: String,
}

fn main() {}
```

#### src/lib.rs

```rust
#![feature(proc_macro)]

extern crate proc_macro;
use proc_macro::TokenStream;

#[proc_macro_derive(AsChangeset, attributes(table_name))]
pub fn derive(input: TokenStream) -> TokenStream {
    for tt in input {
        println!("{:#?}", tt);
    }
    TokenStream::empty()
}
```

**Comment by sgrif** [comment]
The span of the pound token is wrong as well. An error on that span points to the right source code, but has the wrong file/line numbers displayed.

```
error: `treat_none_as_null` must be in the form `treat_none_as_null = "true"`
 --> <macro expansion>:1:1
  |
1 | #[changeset_options(treat_none_as_null("true"))]
  | ^
```

**Comment by alexcrichton** [comment]
This is caused by https://github.com/rust-lang/rust/issues/43081, specifically [this block of code](https://github.com/rust-lang/rust/blob/e0635188ed30e083245482218a0de9237f14ce90/src/libsyntax/parse/token.rs#L797-L800)

**Comment by peterhuene** [comment]
I suspect #52162 is a dupe of this issue?  I'll close.

**Comment by sgrif** [comment]
This has continued to regress. In the most recent nightly, not even the pound token (first token) of the attribute has a meaningful span.

**Comment by peterhuene** [comment]
Indeed, that was the behavior I was observing in my repro (https://github.com/peterhuene/inner-attr-span-repro)

**Comment by peterhuene** [comment]
I've verified the fix in the latest nightly.  Thanks very much!
